### PR TITLE
Fixed dm bug on Anounce command

### LINKED ISSUE #241: Announce failing to send dm announcements
### Describe the bug
Running the k!announce send command to send a message to all users fails with:
AttributeError
'ClientUser' object has no attribute 'create_dm

### To Reproduce
Numbered steps to reproduce the behaviour:
1. Create an empty server
2. Configure KoalaBot to run
3. Send k!announce create
4. Give the announcement a message
5. Send k!announce send

### Expected behaviour
The bot should send the announcement to all users on the server, or with a specified role

### Screenshots
image

### Additional context
None

### PR DESCRIPTION
**Pull request by MoistBiscuits**
## Summary
Fixes bug with k!announce send that was preventing dms from being sent due to users who had dms disabled
closes #241 

## Checklist

- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

<br>

- [x] Have you tested the changes? ([pytest](https://docs.pytest.org/) & [dpytest](https://dpytest.readthedocs.io/))
- [ ] Have you followed [PEP-8](https://www.python.org/dev/peps/pep-0008/) for naming and styling?  
- [ ] Has your code been properly documented with RestructuredText docstrings?
- [ ] Have you added your changes to `CHANGELOG.md` under the `[Unreleased]` heading?
- [ ] If your code added new bot commands, have you updated `documentation.json`?

<br>

- [x] All of this code is your own, or follows the author repo's licence.

### CODE REVIEW

**Comment by MoistBiscuits** [comment]
The bug was caused by attempting to send dm's to bots, which can be of type ClientUser if the bot is a client bot. ClientUser's are read-only so you cannot send dm's to them. Users who have server dm's off were also not handled, so Forbidden exceptions are handled.

**Review by JayDwee — CHANGES_REQUESTED** [review]
Excessive prints should be removed, unless an error or warning has occured to keep the terminal running koala clear.
Also `logging` should be used for this instead of print (see TwitchAlert)

And then you just need to work out why the tests are failing and fix the issue, then request another review from myself and @Kaspiaan (only 1 approval needed though)

**Comment by MoistBiscuits** [comment]
Tests fail if bots aren't allowed to recieve dms since the test cases use bots to check, replaced with eception handling for 3 Exceptions from failling to send a dm. Sorry about the prints, forgot they were left in.

**Comment by JayDwee** [comment]
> Tests fail if bots aren't allowed to recieve dms since the test cases use bots to check, replaced with eception handling for 3 Exceptions from failling to send a dm. Sorry about the prints, forgot they were left in.

If this is the case, make an edit to the tests, or add new tests to show the intended result occurs with and without a bot being sent the message

(I.e. Add tests that would fail in the previous version, and tests that would pass in this version)

**Comment by MoistBiscuits** [comment]
Since I changed the code to not check for bots explicitly, but rather not hang on failed dm's, I don't think new tests are needed. I could re-add bot check if youd like, though I do not think it is neccesary. From testing, code should be fine as is.

**Comment by JayDwee** [comment]
Didn't realise tests were passing, saw checks failed and got confused. Look into the codecov report and increase coverage of tests before we can merge this branch

**Comment by MoistBiscuits** [comment]
Got it, will work on that now

File: KoalaBot.py [hunk 1]
```diff
@@ -151,6 +153,9 @@ def load_all_cogs():
     Loads all cogs in COGS_DIR into the client
     """
     UNRELEASED = []
+    
+    #Auto generate documentation.json
+    generate_doc()
```
Comment (JayDwee) [inline KoalaBot.py:158]: generate doc would be better as its own program, run in GitHub Actions

File: cogs/Announce.py [hunk 2]
```diff
@@ -278,10 +283,17 @@ async def send(self, ctx):
             embed = self.construct_embed(ctx.guild)
             if self.roles[ctx.guild.id]:
                 for receiver in self.get_receivers(ctx.guild.id, ctx.guild.roles):
-                    await receiver.send(embed=embed)
+                    try:
+                        await receiver.send(embed=embed)
+                    except (discord.Forbidden, AttributeError, discord.HTTPException) as e:
+                        logging.error(f'User {receiver.id} cannot recieve dms')
```
Comment (JayDwee) [inline cogs/Announce.py:289]: Add tests for this please

File: cogs/Announce.py [hunk 3]
```diff
@@ -278,10 +283,17 @@ async def send(self, ctx):
...
             else:
                 for receiver in ctx.guild.members:
-                    await receiver.send(embed=embed)
+                    try:
+                        await receiver.send(embed=embed)
+                    except (discord.Forbidden, AttributeError, discord.HTTPException) as e:
+                        logging.error(f'User {receiver.id} cannot recieve dms')
```
Comment (JayDwee) [inline cogs/Announce.py:295]: Add tests for this

File: documentation.json [hunk 4]
```diff
@@ -1,417 +1,418 @@
...
-      },
-      {
-        "command": "vote checkResults",
-        "parameters": ["title"],
-        "description": "Checks the results of a vote without closing it"
-      }
-    ]
-  }
-]
+    {
```
Comment (JayDwee) [inline documentation.json:2]: Should only be indented with 2 spaces, not 4

File: tests/TestDocumentation.json [hunk 5]
```diff
@@ -0,0 +1,418 @@
+[
```
Comment (JayDwee) [inline tests/TestDocumentation.json:1]: What is this file for?

**Review by JayDwee — DISMISSED** [review]
GenerateDoc should be in it's own PR, not here. If you cannot work out how to separate it, please mention GenerateDoc in the PR summary/ description

Also the CHANGELOG.md needs adding to

**Comment by MoistBiscuits** [comment]
Sorry, I had no idea that the generate doc files where being commited, they have their own pr and should not be here. I'll add full test coverage for tests.

File: tests/TestDocumentation.json (hunk 5 above)
Comment (MoistBiscuits) [reply tests/TestDocumentation.json]: Code accidently added for another pr, removed

File: cogs/Announce.py (hunk 3 above)
Comment (MoistBiscuits) [reply cogs/Announce.py:295]: Added test for failed dm sending

File: cogs/Announce.py (hunk 2 above)
Comment (MoistBiscuits) [reply cogs/Announce.py:289]: Added test for sending to a specfic role and for failing dms to a specifc role

File: KoalaBot.py (hunk 1 above)
Comment (MoistBiscuits) [reply KoalaBot.py]: Code accidently added from another pr, removed

**Review by JayDwee — APPROVED** [review]
LGTM 👍 @Kaspiaan double check and merge when ready

**Review by Kaspiaan — APPROVED** [review]
Everything looks good

**Pull request merged**